### PR TITLE
WIP async filter rebase and cleanup of #3300

### PR DESCRIPTION
**Pull request by igorT**

### CODE REVIEW

**Comment by igorT** [comment]
@H1D

File: packages/ember-data/lib/system/record-array-manager.js [hunk 1]
```diff
@@ -121,11 +122,28 @@ export default Ember.Object.extend({
   */
   updateFilterRecordArray: function(array, filter, typeClass, record) {
     var shouldBeInArray = filter(record.getRecord());
-    var recordArrays = this.recordArraysForRecord(record);
+    var key = guidFor(record);
+    if (shouldBeInArray && shouldBeInArray.then) {
+      array.lastFilterPromises[key] = shouldBeInArray;
+      shouldBeInArray.then((result) => {
+        // fighting race conditions here
+        if (shouldBeInArray !== array.lastFilterPromises[key]) {
```
Comment (pangratz) [inline packages/ember-data/lib/system/record-array-manager.js]: Why not excluding the record if the promise rejects?

Comment (H1D) [inline packages/ember-data/lib/system/record-array-manager.js]: @pangratz hmm.. seems like your are right
@igorT something like this needs to be added:
```javascript
shouldBeInArray.catch( => {
  // fighting race conditions here
  if (shouldBeInArray !== array.lastFilterPromises[key]) {
    return;
  }

  delete array.lastFilterPromises[key];
  this._updateFilterRecordArray(array, false, typeClass, record);
});

```

Comment (igorT) [inline packages/ember-data/lib/system/record-array-manager.js]: promise rejecting adds no information, so we take no action.

File: packages/ember-data/lib/system/record-array-manager.js [hunk 2]
```diff
@@ -121,11 +122,28 @@ export default Ember.Object.extend({
...
+      this._updateFilterRecordArray(array, shouldBeInArray, typeClass, record);
+    }
+  },
 
+  _updateFilterRecordArray: function(array, shouldBeInArray, typeClass, record) {
+    var recordArrays = this.recordArraysForRecord(record);
     if (shouldBeInArray) {
       this._addRecordToRecordArray(array, record);
-    } else {
+    } else if (!shouldBeInArray) {
```
Comment (pangratz) [inline packages/ember-data/lib/system/record-array-manager.js]: I think this is a left-over from rebasing. The `else if (!shouldBeInArray)` has been removed as part of cleaning up in [#3167](https://github.com/pangratz/data/blob/b9044fd978c2ff4fc6e6acda1800c57e59351454/packages/ember-data/lib/system/record-array-manager.js#L128)

File: packages/ember-data/lib/system/record-array-manager.js (hunk 1 above)
Comment (pangratz) [inline packages/ember-data/lib/system/record-array-manager.js]: Ah, ok. Thanks for the heads up ...

File: packages/ember-data/lib/system/record-array-manager.js (hunk 2 above)
Comment (sly7-7) [inline packages/ember-data/lib/system/record-array-manager.js]: my bad :disappointed_relieved:

Comment (pangratz) [inline packages/ember-data/lib/system/record-array-manager.js]: Addresses in #3332.

**Comment by bmac** [comment]
The filter api has been moved into an addon. https://github.com/ember-data/ember-data-filter

I'm going to close this pr. Feel free to reopen if you disagree.

**Pull request closed without merging**